avoid specializing strides with DDPOptimizer + inductor

bdhirsh opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/140229

Fixes https://github.com/pytorch/pytorch/issues/139474

The issue was that:

(1) DDPOptimizer has some logic to partition the dynamo graph into buckets, and run AOTAutograd/inductor on each bucket

(2) doing so requires knowing the **exact** strides of the outputs of each subgraph, so we can have example inputs (with correct strides) to each of the later subgraphs to compile with

(3) there is some existing logic to do this today: we have a `fakify_first_call` flag in AOTAutograd that lets you run it with fake tensor inputs (to handle the calling convention changes that AOTAutograd performs at runtime). During this process, we query inductor for the output strides that it compiled with

(4) these outputs strides are stored in the FX graph cache as raw strings of sympy expressions. We have a function, `evaluate_symexpr`, which given the sympy string, and the ShapeEnv's `var_to_val` mapping, will evaluate the sympy string to generate concrete strides

(5) evaluating this expression will specialize on the exact values of any variables in our shape env, however. In DDPOptimizer, we want to know what inductor's stride outputs are symbolically. This requires converting the (string) sympy expression into actual `SymInts` that we can return.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140751



cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @penguinwu @bobrenjc93 @voznesenskym @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

bdhirsh committed:
avoid specializing strides with DDPOptimizer + inductor

[ghstack-poisoned]

eellison commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5151,11 +5151,19 @@ def produce_guards_expression(
-        args = {str(e): val for e, val in self.var_to_val.items()}
+        if return_symints:
+            args = {
+                str(e): SymInt(SymNode(e, self, int, int(val), fx_node=None))
```
I guess this is O(num_output_symints * num_total_symints) ..

bdhirsh replied:
True. This seemed sus to I tried running a small benchmark (function with 100 3D inputs compiled with dynamic=True, P1681754466) with `TORCH_COMPILE_CPROFILE=1`, and:

(1) I don't see `evaluate_symexpr` or `set_tracing_context_output_strides` in the svg

(2) The E2E compile time was 15s - it seems like there is probably other low hanging fruit in a heavily-dynamic microbenchmark to find compile time wins

svg download command: `jf download GG0n2huLD3y3VGsBAAc3IKj-v5g0bsIXAAAz --file "_compile_inner_0_2.svg"`

bdhirsh replied:
(15s was a lot, ended up filing an issue about it here 😛 https://github.com/pytorch/pytorch/issues/140970)

eellison reviewed (commented):
looks good a few questions.. think this was a recent regression with the aot autograd cache changes..

bdhirsh commented on test/distributed/test_dynamo_distributed.py:
```diff
@@ -301,7 +301,10 @@ def _inside_ddp_forward(self):
 
     def forward(self, *inputs, **kwargs):
-        with self._inside_ddp_forward():
+        if not DDP._active_ddp_module:
+            with self._inside_ddp_forward():
+                return self.module.forward(*inputs, **kwargs)
```
When I just re-used this testing infra as-is, dynamo would graph break on the context manager (because it mutates an attribute on the current `nn.Module`) - that might have been fine for other usages of this helper in the tests, but I wanted my test to not have graph breaks, since it makes it easier for me to assert on compile metrics.

My understanding of this test class is that `_active_ddp_module` is a custom attribute in DDP that dynamo hooks into in order to know whether DDP is running, and we can use that attribute to simulate DDP <> dynamo interactions without needing to actual involve distributed code in the test.

I didn't want to fix all of the infra, but I also wanted my test not to have graph breaks, so for my own local usage I set `_inside_ddp_forward()` manually to be true in my test, outside of the compiled region.

If you'd rather I don't tweak the existing testing infra, I can also just tweak my test to not re-use this infra.

bdhirsh commented on torch/_inductor/utils.py:
```diff
@@ -2132,10 +2132,15 @@ def set_tracing_context_output_strides(example_inputs, compiled_graph):
                     tuple(
                         (
-                            shape_env.evaluate_symexpr(e)
+                            shape_env.evaluate_symexpr(
```
Yeah looks like you're right, this must have been regressed from the inductor FX graph cache changes

> Would it be enough to suppress guards here ?

fair question: I think the main issue is that:

(1) prior to the inductor cache, the output strides were actual SymInts - so we could just stash them in the tracing context, grab them in AOTAutograd, and slam them into our output tensors with a call to `as_strided()` while suppressing guards to make sure we didn't specialize

(2) now that we have the inductor cache, the output strides become part of the cache and they are serialized to disk. In the case of an inductor cache hit, we will load them from disk, and `exprs` here is a list of raw strings (they are basically the serialized sympy expressions for each stride).

So at the very least, we need to convert the raw sympy strings back into SymInts to use them in AOTAutograd. I found that `ShapeEnv.evaluate_symexpr` was the existing code for doing this: but it would also replace each symbol with the current hint, instead of replacing it with the raw `sympy.Symbol` that it mapped to.

eellison replied:
cc @jamesjwu just to be aware.

bdhirsh commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5151,11 +5151,19 @@ def produce_guards_expression(
 
-    def evaluate_symexpr(self, code: str) -> Union[int, float, bool]:
+    def evaluate_symexpr(
+        self, code: str, *, return_symints: bool
```
yup agreed - I'll split it into a separate function

bdhirsh committed:
Update on "avoid specializing strides with DDPOptimizer + inductor"

Fixes https://github.com/pytorch/pytorch/issues/140229

The issue was that:

(1) DDPOptimizer has some logic to partition the dynamo graph into buckets, and run AOTAutograd/inductor on each bucket

(2) doing so requires knowing the **exact** strides of the outputs of each subgraph, so we can have example inputs (with correct strides) to each of the later subgraphs to compile with

(3) there is some existing logic to do this today: we have a `fakify_first_call` flag in AOTAutograd that lets you run it with fake tensor inputs (to handle the calling convention changes that AOTAutograd performs at runtime). During this process, we query inductor for the output strides that it compiled with

(4) these outputs strides are stored in the FX graph cache as raw strings of sympy expressions. We have a function, `evaluate_symexpr`, which given the sympy string, and the ShapeEnv's `var_to_val` mapping, will evaluate the sympy string to generate concrete strides

(5) evaluating this expression will specialize on the exact values of any variables in our shape env, however. In DDPOptimizer, we want to know what inductor's stride outputs are symbolically. This requires converting the (string) sympy expression into actual `SymInts` that we can return.




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv penguinwu bobrenjc93 voznesenskym Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "avoid specializing strides with DDPOptimizer + inductor"

Fixes https://github.com/pytorch/pytorch/issues/140229

Fixes https://github.com/pytorch/pytorch/issues/139474

The issue was that:

(1) DDPOptimizer has some logic to partition the dynamo graph into buckets, and run AOTAutograd/inductor on each bucket

(2) doing so requires knowing the **exact** strides of the outputs of each subgraph, so we can have example inputs (with correct strides) to each of the later subgraphs to compile with

(3) there is some existing logic to do this today: we have a `fakify_first_call` flag in AOTAutograd that lets you run it with fake tensor inputs (to handle the calling convention changes that AOTAutograd performs at runtime). During this process, we query inductor for the output strides that it compiled with

(4) these outputs strides are stored in the FX graph cache as raw strings of sympy expressions. We have a function, `evaluate_symexpr`, which given the sympy string, and the ShapeEnv's `var_to_val` mapping, will evaluate the sympy string to generate concrete strides

(5) evaluating this expression will specialize on the exact values of any variables in our shape env, however. In DDPOptimizer, we want to know what inductor's stride outputs are symbolically. This requires converting the (string) sympy expression into actual `SymInts` that we can return.




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv penguinwu bobrenjc93 voznesenskym Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

bdhirsh committed:
Update on "avoid specializing strides with DDPOptimizer + inductor"

Fixes https://github.com/pytorch/pytorch/issues/140229

Fixes https://github.com/pytorch/pytorch/issues/139474

The issue was that:

(1) DDPOptimizer has some logic to partition the dynamo graph into buckets, and run AOTAutograd/inductor on each bucket

(2) doing so requires knowing the **exact** strides of the outputs of each subgraph, so we can have example inputs (with correct strides) to each of the later subgraphs to compile with

(3) there is some existing logic to do this today: we have a `fakify_first_call` flag in AOTAutograd that lets you run it with fake tensor inputs (to handle the calling convention changes that AOTAutograd performs at runtime). During this process, we query inductor for the output strides that it compiled with

(4) these outputs strides are stored in the FX graph cache as raw strings of sympy expressions. We have a function, `evaluate_symexpr`, which given the sympy string, and the ShapeEnv's `var_to_val` mapping, will evaluate the sympy string to generate concrete strides

(5) evaluating this expression will specialize on the exact values of any variables in our shape env, however. In DDPOptimizer, we want to know what inductor's stride outputs are symbolically. This requires converting the (string) sympy expression into actual `SymInts` that we can return.




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv penguinwu bobrenjc93 voznesenskym Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
